### ISSUE TITLE
Add to chrome extensions

### ISSUE THREAD

**Issue by etcimon** [jcbrand/converse.js#153]
It would be nice to have this as a chrome extension. There are currently no XMPP extensions for chrome and it would be the only available application for Chrome OS

**Comment by jcbrand** [comment]
This weekend I started working on making a Chrome extension based on converse.js. However, I've realized that the kind of extension that I ideally would like implement is not possible.

Ideally, I would like the extension to be independent of any particular XMPP server or any server-side web application. That way, you can log in with your existing XMPP credentials from where-ever.

Chrome doesn't allow a secure way to store user data, so storing user credentials via [chrome.storage](https://stackoverflow.com/questions/3937000/chrome-extension-accessing-localstorage-in-content-script) is not a good idea. The `keepalive` feature of converse.js means that you need to log in only once at the start, and from then on your session is maintained across page loads. However, this is only per tab, or at least per domain.

So whenever you switch tabs, converse.js needs to be logged in anew. If I could securely store the XMPP credentials somewhere, then this could be done automatically and wouldn't be a problem. But because credentials cannot be stored securely in your browser or in chrome.storage, it means that the user needs to log in again and again for each individual tab.

Here are some workarounds, I've been able to come up:

- **Require users to register new accounts on a web-app tied to an XMPP server under my control, so that it can do prebinding for them.** This means user's can't use their own XMPP accounts but will have to register new ones from a new server. Not what I originally would have liked.
- **Store user credentials somewhere on a web application that also does prebinding.** As a user, this means my credentials need to be stored in plaintext on someone else's server. Not something I'd agree to. Also, this means I as the administrator of the web application have the responsibility of keeping safe the plaintext XMPP credentials of users. Not something I'd like to deal with.
- **To hell with it, store the password encrypted in localStorage.** This requires Javascript crypto in the Chrome extension and in JS code that gets injected into the page and because the code the encrypts/decrypts the password is available in the browser, the password can be decrypted.

So for now I'm putting this on ice again. I'm all ears if someone has any suggestions on how this problem could be overcome.

**Comment by etcimon** [comment]
As far as I'm concerned, a "Secure" cookie containing the password could do, because 

1) it's domain specific
2) it was meant even for encrypted banking sessions
3) you can access it from javascript if you don't set it HTTPOnly
4) it can be session-based

**Comment by jcbrand** [comment]
This won't work for the following reasons:

* We need the password in plaintext. Banking websites definitely do not store plaintext passwords in cookies, they would store hashes, which is of no use to us.
* Since it's session-based, this would mean a special cookie for every website which the user views (and on which the extension is active) and this cookie would be included in every request which is made to that website's backend.
* The fact that you can access it from JS is a big drawback because all JS on a page (which we don't control) can read the cookie.
